Extension loading errors breaking external extension load

### Bug description

If one extension has misconfigured `main` script, or that script not present on filesystem for some reasons, it breaks rest of extension to load.

The code responsible for extension loading is there:
https://github.com/containers/podman-desktop/blob/23ccff4cf6bd778023d8a3543c00ed41d51a97c1/packages/main/src/plugin/extension-loader.ts#L226-L227

So if any error appears during the extension load, there are no code to handle that error, and whole sequence of loading extension is broken, so rest of extension will not load.



### Operating system

 macOS 13.2.1 Intel

### Version

0.14.1

### Steps to reproduce

Just break one of the builtin extension, by changing in package.json `main` field, to point to non existing file.
And start application, you will see that extensions that should be loaded after extension that you broke, will not load. 

### Relevant log output

```shell
(node:76301) UnhandledPromiseRejectionWarning: Error: Cannot find module '/Users/yevhen/work/redhat/podman-desktop/extensions/podman/dist/extensionaqawww.js'
Require stack:
- /Users/yevhen/work/redhat/podman-desktop/packages/main/dist/index.cjs
- /Users/yevhen/work/redhat/podman-desktop/node_modules/electron/dist/Electron.app/Contents/Resources/default_app.asar/main.js
- 
    at Module._resolveFilename (node:internal/modules/cjs/loader:1054:15)
    at n._resolveFilename (node:electron/js2c/browser_init:2:109877)
    at Module._load (node:internal/modules/cjs/loader:900:27)
    at f._load (node:electron/js2c/asar_bundle:2:13330)
    at module2._load (/Users/yevhen/work/redhat/podman-desktop/packages/main/dist/index.cjs:37643:31)
    at Module.require (node:internal/modules/cjs/loader:1120:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at ExtensionLoader.loadRuntime (/Users/yevhen/work/redhat/podman-desktop/packages/main/dist/index.cjs:38090:12)
    at ExtensionLoader.loadExtension (/Users/yevhen/work/redhat/podman-desktop/packages/main/dist/index.cjs:37797:26)
    at async Promise.all (index 5)
```
```


### Additional context

_No response_

--- a/packages/main/src/plugin/api/extension-info.ts
+++ b/packages/main/src/plugin/api/extension-info.ts
@@ -16,6 +16,11 @@
  * SPDX-License-Identifier: Apache-2.0
  ***********************************************************************/
 
+export interface ExtensionError {
+  message: string;
+  stack?: string;
+}
+
 export interface ExtensionInfo {
   id: string;
   name: string;
@@ -25,5 +30,6 @@ export interface ExtensionInfo {
   removable: boolean;
   version: string;
   state: string;
+  error?: ExtensionError;
   path: string;
 }

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -96,7 +96,8 @@ const inputQuickPickRegistry: InputQuickPickRegistry = {} as unknown as InputQui
 
 const authenticationProviderRegistry: AuthenticationImpl = {} as unknown as AuthenticationImpl;
 
-const telemetry: Telemetry = { track: vi.fn() } as unknown as Telemetry;
+const telemetryTrackMock = vi.fn();
+const telemetry: Telemetry = { track: telemetryTrackMock } as unknown as Telemetry;
 
 /* eslint-disable @typescript-eslint/no-empty-function */
 beforeAll(() => {
@@ -123,6 +124,7 @@ beforeAll(() => {
 });
 
 beforeEach(() => {
+  telemetryTrackMock.mockImplementation(() => Promise.resolve());
   vi.clearAllMocks();
 });
 

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -53,6 +53,10 @@ class TestExtensionLoader extends ExtensionLoader {
   setWatchTimeout(timeout: number): void {
     this.watchTimeout = timeout;
   }
+
+  getExtensionState() {
+    return this.extensionState;
+  }
 }
 
 let extensionLoader: TestExtensionLoader;
@@ -67,7 +71,7 @@ const configurationRegistry: ConfigurationRegistry = {} as unknown as Configurat
 
 const imageRegistry: ImageRegistry = {} as unknown as ImageRegistry;
 
-const apiSender: ApiSenderType = {} as unknown as ApiSenderType;
+const apiSender: ApiSenderType = { send: vi.fn() } as unknown as ApiSenderType;
 
 const trayMenuRegistry: TrayMenuRegistry = {} as unknown as TrayMenuRegistry;
 
@@ -229,4 +233,24 @@ test('Should load file from watching scanning folder', async () => {
 
   // expect to load only one file (other are invalid files/folder)
   expect(loadPackagedFileMock).toBeCalledWith(path.resolve(rootedFakeDirectory, 'watch.cdix'));
+});
+
+test('Verify extension error leads to errored state', async () => {
+  const id = 'extension.id';
+  await extensionLoader.activateExtension(
+    {
+      id: id,
+      path: 'dummy',
+      api: undefined,
+      mainPath: '',
+      removable: false,
+      manifest: {},
+    },
+    {
+      activate: () => {
+        throw Error('Failed');
+      },
+    },
+  );
+  expect(extensionLoader.getExtensionState().get(id)).toBe('errored');
 });

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -40,6 +40,7 @@ import type { ApiSenderType } from './api';
 import type { AuthenticationImpl } from './authentication';
 import type { MessageBox } from './message-box';
 import type { Telemetry } from './telemetry/telemetry';
+import type * as containerDesktopAPI from '@podman-desktop/api';
 
 class TestExtensionLoader extends ExtensionLoader {
   public async setupScanningDirectory(): Promise<void> {
@@ -241,7 +242,7 @@ test('Verify extension error leads to failed state', async () => {
     {
       id: id,
       path: 'dummy',
-      api: {},
+      api: {} as typeof containerDesktopAPI,
       mainPath: '',
       removable: false,
       manifest: {},

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -241,7 +241,7 @@ test('Verify extension error leads to failed state', async () => {
     {
       id: id,
       path: 'dummy',
-      api: undefined,
+      api: {},
       mainPath: '',
       removable: false,
       manifest: {},

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -235,7 +235,7 @@ test('Should load file from watching scanning folder', async () => {
   expect(loadPackagedFileMock).toBeCalledWith(path.resolve(rootedFakeDirectory, 'watch.cdix'));
 });
 
-test('Verify extension error leads to errored state', async () => {
+test('Verify extension error leads to failed state', async () => {
   const id = 'extension.id';
   await extensionLoader.activateExtension(
     {
@@ -252,5 +252,5 @@ test('Verify extension error leads to errored state', async () => {
       },
     },
   );
-  expect(extensionLoader.getExtensionState().get(id)).toBe('errored');
+  expect(extensionLoader.getExtensionState().get(id)).toBe('failed');
 });

--- a/packages/main/src/plugin/extension-loader.spec.ts
+++ b/packages/main/src/plugin/extension-loader.spec.ts
@@ -95,7 +95,7 @@ const inputQuickPickRegistry: InputQuickPickRegistry = {} as unknown as InputQui
 
 const authenticationProviderRegistry: AuthenticationImpl = {} as unknown as AuthenticationImpl;
 
-const telemetry: Telemetry = {} as unknown as Telemetry;
+const telemetry: Telemetry = { track: vi.fn() } as unknown as Telemetry;
 
 /* eslint-disable @typescript-eslint/no-empty-function */
 beforeAll(() => {

--- a/packages/main/src/plugin/extension-loader.ts
+++ b/packages/main/src/plugin/extension-loader.ts
@@ -21,7 +21,7 @@ import * as path from 'path';
 import * as os from 'os';
 import * as fs from 'fs';
 import type { CommandRegistry } from './command-registry';
-import type { ExtensionInfo } from './api/extension-info';
+import type { ExtensionError, ExtensionInfo } from './api/extension-info';
 import * as zipper from 'zip-local';
 import type { TrayMenuRegistry } from './tray-menu-registry';
 import { Disposable } from './types/disposable';
@@ -90,6 +90,7 @@ export class ExtensionLoader {
   private watcherExtensions = new Map<string, containerDesktopAPI.FileSystemWatcher>();
   private reloadInProgressExtensions = new Map<string, boolean>();
   protected extensionState = new Map<string, string>();
+  protected extensionStateErrors = new Map<string, unknown>();
 
   protected watchTimeout = 1000;
 
@@ -121,6 +122,18 @@ export class ExtensionLoader {
     private telemetry: Telemetry,
   ) {}
 
+  mapError(err: unknown): ExtensionError | undefined {
+    if (err) {
+      if (err instanceof Error) {
+        return { message: err.message, stack: err.stack };
+      } else {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        return { message: (err as any).toString() };
+      }
+    }
+    return undefined;
+  }
+
   async listExtensions(): Promise<ExtensionInfo[]> {
     return Array.from(this.analyzedExtensions.values()).map(extension => ({
       name: extension.manifest.name,
@@ -129,6 +142,7 @@ export class ExtensionLoader {
       version: extension.manifest.version,
       publisher: extension.manifest.publisher,
       state: this.extensionState.get(extension.id) || 'stopped',
+      error: this.mapError(this.extensionStateErrors.get(extension.id)),
       id: extension.id,
       path: extension.path,
       removable: extension.removable,
@@ -774,6 +788,7 @@ export class ExtensionLoader {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   async activateExtension(extension: AnalyzedExtension, extensionMain: any): Promise<void> {
     this.extensionState.set(extension.id, 'starting');
+    this.extensionStateErrors.delete(extension.id);
     this.apiSender.send('extension-starting', {});
 
     const subscriptions: containerDesktopAPI.Disposable[] = [];
@@ -805,6 +820,7 @@ export class ExtensionLoader {
     } catch (err) {
       console.log(`Activation extension ${extension.id} failed error:${err}`);
       this.extensionState.set(extension.id, 'failed');
+      this.extensionStateErrors.set(extension.id, err);
     }
   }
 

--- a/packages/main/src/plugin/extension-loader.ts
+++ b/packages/main/src/plugin/extension-loader.ts
@@ -384,25 +384,40 @@ export class ExtensionLoader {
     }
 
     this.analyzedExtensions.set(extension.id, extension);
+    this.extensionState.delete(extension.id);
+    this.extensionStateErrors.delete(extension.id);
 
-    // in development mode, watch if the extension is updated and reload it
-    if (import.meta.env.DEV && !this.watcherExtensions.has(extension.id)) {
-      const extensionWatcher = this.fileSystemMonitoring.createFileSystemWatcher(extensionPath);
-      extensionWatcher.onDidChange(async () => {
-        // wait 1 second before trying to reload the extension
-        // this is to avoid reloading the extension while it is still being updated
-        setTimeout(() => {
-          this.reloadExtension(extension, removable).catch((error: unknown) =>
-            console.error('error while reloading extension', error),
-          );
-        }, 1000);
+    const telemetryOptions = { extensionId: extension.id };
+
+    try {
+      // in development mode, watch if the extension is updated and reload it
+      if (import.meta.env.DEV && !this.watcherExtensions.has(extension.id)) {
+        const extensionWatcher = this.fileSystemMonitoring.createFileSystemWatcher(extensionPath);
+        extensionWatcher.onDidChange(async () => {
+          // wait 1 second before trying to reload the extension
+          // this is to avoid reloading the extension while it is still being updated
+          setTimeout(() => {
+            this.reloadExtension(extension, removable).catch((error: unknown) =>
+              console.error('error while reloading extension', error),
+            );
+          }, 1000);
+        });
+        this.watcherExtensions.set(extension.id, extensionWatcher);
+      }
+
+      const runtime = this.loadRuntime(extension);
+
+      await this.activateExtension(extension, runtime);
+    } catch (err) {
+      this.extensionState.set(extension.id, 'failed');
+      this.extensionStateErrors.set(extension.id, err);
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (telemetryOptions as any).error = err;
+    } finally {
+      this.telemetry.track('loadExtension', telemetryOptions).catch((error: unknown) => {
+        console.error('error while tracking loadExtension telemetry', error);
       });
-      this.watcherExtensions.set(extension.id, extensionWatcher);
     }
-
-    const runtime = this.loadRuntime(extension);
-
-    await this.activateExtension(extension, runtime);
   }
 
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
@@ -805,7 +820,7 @@ export class ExtensionLoader {
     const telemetryOptions = { extensionId: extension.id };
     try {
       if (typeof extensionMain['activate'] === 'function') {
-        // return exports
+        // it returns exports
         console.log(`Activating extension (${extension.id})`);
         await extensionMain['activate'].apply(undefined, [extensionContext]);
         console.log(`Activation extension (${extension.id}) ended`);
@@ -826,7 +841,9 @@ export class ExtensionLoader {
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       (telemetryOptions as any).error = err;
     } finally {
-      this.telemetry.track('activateExtension', telemetryOptions);
+      this.telemetry
+        .track('activateExtension', telemetryOptions)
+        .catch((error: unknown) => console.log(`Failed to track activateExtension telemetry event. Error: ${error}`));
     }
   }
 
@@ -870,7 +887,9 @@ export class ExtensionLoader {
     this.activatedExtensions.delete(extensionId);
     this.extensionState.set(extension.id, 'stopped');
     this.apiSender.send('extension-stopped');
-    this.telemetry.track('deactivateExtension', telemetryOptions);
+    this.telemetry
+      .track('deactivateExtension', telemetryOptions)
+      .catch((error: unknown) => console.log(`Failed to track deactivateExtension telemetry event. Error: ${error}`));
   }
 
   async stopAllExtensions(): Promise<void> {

--- a/packages/main/src/plugin/extension-loader.ts
+++ b/packages/main/src/plugin/extension-loader.ts
@@ -89,7 +89,7 @@ export class ExtensionLoader {
   private analyzedExtensions = new Map<string, AnalyzedExtension>();
   private watcherExtensions = new Map<string, containerDesktopAPI.FileSystemWatcher>();
   private reloadInProgressExtensions = new Map<string, boolean>();
-  private extensionState = new Map<string, string>();
+  protected extensionState = new Map<string, string>();
 
   protected watchTimeout = 1000;
 
@@ -786,21 +786,26 @@ export class ExtensionLoader {
     if (typeof extensionMain['deactivate'] === 'function') {
       deactivateFunction = extensionMain['deactivate'];
     }
-    if (typeof extensionMain['activate'] === 'function') {
-      // activate the extension
-      console.log(`Activating extension (${extension.id})`);
-      await extensionMain['activate'].apply(undefined, [extensionContext]);
-      console.log(`Activation extension (${extension.id}) ended`);
+    try {
+      if (typeof extensionMain['activate'] === 'function') {
+        // return exports
+        console.log(`Activating extension (${extension.id})`);
+        await extensionMain['activate'].apply(undefined, [extensionContext]);
+        console.log(`Activation extension (${extension.id}) ended`);
+      }
+      const id = extension.id;
+      const activatedExtension: ActivatedExtension = {
+        id,
+        deactivateFunction,
+        extensionContext,
+      };
+      this.activatedExtensions.set(extension.id, activatedExtension);
+      this.extensionState.set(extension.id, 'started');
+      this.apiSender.send('extension-started');
+    } catch (err) {
+      console.log(`Activation extension ${extension.id} errored error:${err}`);
+      this.extensionState.set(extension.id, 'errored');
     }
-    const id = extension.id;
-    const activatedExtension: ActivatedExtension = {
-      id,
-      deactivateFunction,
-      extensionContext,
-    };
-    this.activatedExtensions.set(extension.id, activatedExtension);
-    this.extensionState.set(extension.id, 'started');
-    this.apiSender.send('extension-started');
   }
 
   async deactivateExtension(extensionId: string): Promise<void> {
@@ -813,7 +818,11 @@ export class ExtensionLoader {
     this.apiSender.send('extension-stopping');
 
     if (extension.deactivateFunction) {
-      await extension.deactivateFunction();
+      try {
+        await extension.deactivateFunction();
+      } catch (err) {
+        console.log(`Deactivation extension ${extension.id} errored error:${err}`);
+      }
     }
 
     // dispose subscriptions

--- a/packages/main/src/plugin/extension-loader.ts
+++ b/packages/main/src/plugin/extension-loader.ts
@@ -803,8 +803,8 @@ export class ExtensionLoader {
       this.extensionState.set(extension.id, 'started');
       this.apiSender.send('extension-started');
     } catch (err) {
-      console.log(`Activation extension ${extension.id} errored error:${err}`);
-      this.extensionState.set(extension.id, 'errored');
+      console.log(`Activation extension ${extension.id} failed error:${err}`);
+      this.extensionState.set(extension.id, 'failed');
     }
   }
 
@@ -821,7 +821,7 @@ export class ExtensionLoader {
       try {
         await extension.deactivateFunction();
       } catch (err) {
-        console.log(`Deactivation extension ${extension.id} errored error:${err}`);
+        console.log(`Deactivation extension ${extension.id} failed error:${err}`);
       }
     }
 

--- a/packages/main/src/plugin/extension-loader.ts
+++ b/packages/main/src/plugin/extension-loader.ts
@@ -801,6 +801,8 @@ export class ExtensionLoader {
     if (typeof extensionMain['deactivate'] === 'function') {
       deactivateFunction = extensionMain['deactivate'];
     }
+
+    const telemetryOptions = { extensionId: extension.id };
     try {
       if (typeof extensionMain['activate'] === 'function') {
         // return exports
@@ -821,6 +823,10 @@ export class ExtensionLoader {
       console.log(`Activation extension ${extension.id} failed error:${err}`);
       this.extensionState.set(extension.id, 'failed');
       this.extensionStateErrors.set(extension.id, err);
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (telemetryOptions as any).error = err;
+    } finally {
+      this.telemetry.track('activateExtension', telemetryOptions);
     }
   }
 
@@ -830,6 +836,8 @@ export class ExtensionLoader {
       return;
     }
 
+    const telemetryOptions = { extensionId: extension.id };
+
     this.extensionState.set(extension.id, 'stopping');
     this.apiSender.send('extension-stopping');
 
@@ -838,6 +846,8 @@ export class ExtensionLoader {
         await extension.deactivateFunction();
       } catch (err) {
         console.log(`Deactivation extension ${extension.id} failed error:${err}`);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (telemetryOptions as any).error = err;
       }
     }
 
@@ -860,6 +870,7 @@ export class ExtensionLoader {
     this.activatedExtensions.delete(extensionId);
     this.extensionState.set(extension.id, 'stopped');
     this.apiSender.send('extension-stopped');
+    this.telemetry.track('deactivateExtension', telemetryOptions);
   }
 
   async stopAllExtensions(): Promise<void> {

--- a/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
+++ b/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
@@ -39,7 +39,7 @@ async function removeExtension() {
           <!-- start is enabled only when stopped -->
           <div class="px-2 text-sm italic text-gray-700">
             <button
-              disabled="{extensionInfo.state !== 'stopped'}"
+              disabled="{extensionInfo.state !== 'stopped' && extensionInfo.state !== 'failed'}"
               on:click="{() => startExtension()}"
               class="pf-c-button pf-m-primary"
               type="button">
@@ -68,7 +68,7 @@ async function removeExtension() {
           {#if extensionInfo.removable}
             <div class="px-2 text-sm italic text-gray-700">
               <button
-                disabled="{extensionInfo.state !== 'stopped'}"
+                disabled="{extensionInfo.state !== 'stopped' && extensionInfo.state !== 'failed'}"
                 on:click="{() => removeExtension()}"
                 class="pf-c-button pf-m-primary"
                 type="button">

--- a/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
+++ b/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
@@ -82,6 +82,15 @@ async function removeExtension() {
             <div class="text-gray-900 items-center px-2 text-sm">Default extension, cannot be removed</div>
           {/if}
         </div>
+        {#if extensionInfo.error}
+          <div class="flex flex-col">
+            <div class="py-2">Extension error: {extensionInfo.error.message}</div>
+            {#if extensionInfo.error.stack}
+              <div class="py-2">Stack trace</div>
+              <div class="py-2">{extensionInfo.error.stack}</div>
+            {/if}
+          </div>
+        {/if}
       </Route>
     {/if}
   </div></SettingsPage>

--- a/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
+++ b/packages/renderer/src/lib/preferences/PreferencesExtensionRendering.svelte
@@ -36,7 +36,7 @@ async function removeExtension() {
         </div>
 
         <div class="py-2 flex flex-row items-center">
-          <!-- start is enabled only when stopped -->
+          <!-- start is enabled only when stopped or failed -->
           <div class="px-2 text-sm italic text-gray-700">
             <button
               disabled="{extensionInfo.state !== 'stopped' && extensionInfo.state !== 'failed'}"
@@ -64,7 +64,7 @@ async function removeExtension() {
             </button>
           </div>
 
-          <!-- delete is enabled only when stopped -->
+          <!-- delete is enabled only when stopped or failed -->
           {#if extensionInfo.removable}
             <div class="px-2 text-sm italic text-gray-700">
               <button

--- a/packages/renderer/src/lib/ui/ConnectionStatus.svelte
+++ b/packages/renderer/src/lib/ui/ConnectionStatus.svelte
@@ -42,6 +42,14 @@ const statusesStyle = new Map<string, connectionStatusStyle>([
       label: 'STOPPING',
     },
   ],
+  [
+    'failed',
+    {
+      bgColor: 'bg-red-500',
+      txtColor: 'text-red-500',
+      label: 'FAILED',
+    },
+  ],
 ]);
 $: statusStyle = statusesStyle.get(status) || {
   bgColor: 'bg-gray-900',

